performance/tests.rs

```diff
@@ -0,0 +1,638 @@
+            "-t",
+            &format!("{}", test_time),
+        ]);
+        // For measuring the guest transmit throughput (as a sender),
+        // use reverse mode of the iperf3 client on the host
+        if !receive {
+            cmd.args(&["-R"]);
+        }
+        let client = cmd
+            .stderr(Stdio::piped())
+            .stdout(Stdio::piped())
+            .spawn()
+            .map_err(Error::Spawn)?;
+
+        clients.push(client);
+    }
+
+    let mut err: Option<Error> = None;
+    let mut bps = Vec::new();
+    let mut failed = false;
+    for c in clients {
+        let mut c = c;
+        if let Err(e) = child_wait_timeout(&mut c, test_time as u64 + 5) {
+            err = Some(Error::WaitTimeout(e));
+            failed = true;
+        }
+
+        if !failed {
+            // Safe to unwrap as we know the child has terminated succesffully
+            let output = c.wait_with_output().unwrap();
+            bps.push(parse_iperf3_output(&output.stdout, receive)?);
+        } else {
+            let _ = c.kill();
+            let output = c.wait_with_output().unwrap();
+            println!(
+                "=============== Client output [Error] ===============\n\n{}\n\n===========end============\n\n",
+                String::from_utf8_lossy(&output.stdout)
+            );
+        }
+    }
+
+    if let Some(e) = err {
+        Err(e)
+    } else {
+        Ok(bps.iter().sum())
+    }
+}
+
+pub fn performance_net_throughput(control: &PerformanceTestControl) -> f64 {
+    let test_time = control.test_time;
+    let queue_pairs = control.queue_num.unwrap();
+    let queue_size = control.queue_size.unwrap();
+    let rx = control.net_rx.unwrap();
+
+    let focal = UbuntuDiskConfig::new(FOCAL_IMAGE_NAME.to_string());
+    let guest = Guest::new(Box::new(focal));
+
+    let net_params = format!(
+        "tap=,mac={},ip={},mask=255.255.255.0,num_queues={},queue_size={}",
+        guest.network.guest_mac,
+        guest.network.host_ip,
+        queue_pairs * 2,
+        queue_size,
+    );
+
+    let mut child = GuestCommand::new(&guest)
+        .args(&["--cpus", &format!("boot={}", queue_pairs * 2)])
+        .args(&["--memory", "size=4G"])
+        .args(&["--kernel", direct_kernel_boot_path().to_str().unwrap()])
+        .args(&["--cmdline", DIRECT_KERNEL_BOOT_CMDLINE])
+        .default_disks()
+        .args(&["--net", net_params.as_str()])
+        .capture_output()
+        .spawn()
+        .unwrap();
+
+    let r = std::panic::catch_unwind(|| {
+        guest.wait_vm_boot(None).unwrap();
+        measure_virtio_net_throughput(test_time, queue_pairs, &guest, rx).unwrap()
+    });
+
+    let _ = child.kill();
+    let output = child.wait_with_output().unwrap();
+
+    match r {
+        Ok(r) => r,
+        Err(e) => {
+            handle_child_output(Err(e), &output);
+            panic!("test failed!");
+        }
+    }
+}
+
+fn parse_ethr_latency_output(output: &[u8]) -> Result<Vec<f64>, Error> {
```

Comment:
I think you could have unit tests for the `parse_` functions that take reference data and check the values extracted make sense.